### ISSUE TITLE
Datashape type string and Impala 

### ISSUE THREAD

**Issue by fkaufer** [blaze/blaze#853]
Datashape type `string` is always converted to SQL types `TEXT` and `VARCHAR`, which are not provided by Impala, but only `STRING`.

For a simple example I derived the following snippet from
https://github.com/mrocklin/blaze-scipy-2014/blob/master/01-computation.ipynb

``` python
cities_list = [(1, 'NYC'),
               (1, 'Boston'),
               (2, 'LA'),
               (3, 'Beijing'),
               (3, 'Shanghai'),
               (4, 'Moscow'),
               (5, 'Paris'),
               (5, 'Berlin')]

cities_schema = "{id: ?int, city: ?string}"

into(SQL('impala://...', 'cities', schema=cities_schema), cities_list)
```

Execution results in 

```
DBAPIError: (HiveServer2Error) AnalysisException: Syntax error in line 4:
    city TEXT
      ^
Encountered: IDENTIFIER
Expected: BIGINT, BINARY, BOOLEAN, CHAR, DATE, DATETIME, DECIMAL, REAL, FLOAT, INTEGER, SMALLINT, STRING, TIMESTAMP, TINYINT
```

Workaround by fixing length (`"{id: ?int, city: ?string[50]}"`) doesn't work because `VARCHAR` is also not supported. 

I guess, [`dshape_to_alchemy()`](https://github.com/ContinuumIO/blaze/blob/0.6.6/blaze/data/sql.py#L95) needs a bit more differentiation based on the SQL dialect.

As an alternative I implemented the following workarkound in impyla's `impala.sqlalchemy`.

``` python
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.types import Text  

@compiles(Text, "impala")
def compile_text_impala(type_, compiler, **kw):
    return "STRING"    
```

It works, but I'm not sure it's the right way to do it. If at all, Impala (i.e. the server) could allow `TEXT` and `VARCHAR` as aliases, but Impyla should probably not auto-adapt.

CC @laserson

**Comment by cpcloud** [comment]
I think this should be pushed to `impyla` and sqlalchemy's type compiler rather than increasing the complexity of `dshape_to_alchemy`. @fkaufer did you submit your fix upstream to impyla?

**Comment by cpcloud** [comment]
PyHive is an example of a SQLAlchemy dialect which does this:

https://github.com/dropbox/PyHive/blob/master/pyhive/sqlalchemy_hive.py#L126-L158

**Comment by fkaufer** [comment]
> did you submit your fix upstream to impyla?

No, because I think it's controversial and personally I thought black magic stuff (abstractions, automatic conversions etc) is rather a job for blaze/odo whereas a database client interface should be close to the system's SQL datatypes and only provide a basic level of abstraction. When you use a database client directly you might want to know about missing/incompatible types and potentially related issues, when you use odo you expect things to work and don't want to deal with idiosyncratic data type issues.

But admittedly, you can also argue in favor of the PyHive solution.

**Comment by cpcloud** [comment]
@fkaufer

> No, because I think it's controversial

What's controversial? The fix?

> ... whereas a database client interface should be close to the system's SQL datatypes and only provide a basic level of abstraction

One of the goals of SQLAlchemy (as I understand it) is to abstract away the idiosyncrasies of all the different RDMBSs. It has explicit mechanisms for doing such things on a per dialect basis.

> When you use a database client directly you might want to know about missing/incompatible types and potentially related issues, when you use odo you expect things to work and don't want to deal with idiosyncratic data type issues

I guess we're of two minds here. I think that sqlalchemy should "just work" in this case (and it usually does). I personally don't care that certain types like `varchar` or `char` are incompatible with impala. With those types, I usually just want a string column. If I cared about faithfully representing those types in my database, then I probably wouldn't use Impala in the first place.

This fix would be likely be trivial to implement in impyla, and it wouldn't increase the code complexity.

I also think that it is the job of the dialect to map sqlalchemy types to actual db types.

There's even a `TypeCompiler` class in SQLAlchemy that can be subclassed to override type compilation behavior.

The only time odo should care about the sqldialect is when i want to choose one to pass to sqlalchemy.

In any event, do you mind if I submit your fix upstream to impyla?

**Comment by fkaufer** [comment]
> What's controversial? The fix?

I would only consider it a fix for odo, but at most a convenience feature for impyla which could turn into a gotcha, hence controversial.

By example: Impala has [VARCHAR support since 2.0/CDH5.2](http://www.cloudera.com/content/cloudera/en/documentation/core/latest/topics/impala_varchar.html). If impyla had compiled VARCHAR to STRING (which was obviously easy to implement) it would now have to break this behavior to support the new language feature. Even if you would not mind this particular disruption I hope you agree that it is controversial and a legitimate approach to throw "not (yet) supported  type exceptions" leaving it to downstream applications dealing explicitly with that.

>> ... whereas a database client interface should be close to the system's SQL datatypes and only provide a basic level of abstraction

> One of the goals of SQLAlchemy (as I understand it) is to abstract away the idiosyncrasies of all the different RDMBSs. It has explicit mechanisms for doing such things on a per dialect basis.

Sure, but the crucial question is: what are syntactical idiosyncrasies and what meaningful differences? The answer might differ depending on the use case, here general-purpose database client vs. high-level data integration/conversion. SQLAlchemy is a flexible tool and can be used by low- and high-level applications and for varying degrees of abstraction. In other words: you can use SQLAlchemy to create both a generic/flexible and a specific/rigor dialect for one and the same database system. You can also map every (not supported) type to BLOB.

> I guess we're of two minds here ... I personally don't care that certain types like varchar or char are incompatible with impala. ... With those types, I usually just want a string column. 

Actually, in this regard we are on the same page and therefore I highly appreciate your work on blaze/odo, taking the burden from me to deal with these subtleties when I don't have/want to. Regarding my current personal use cases, I wouldn't not mind if impyla's sqlalchemy dialect compiles TEXT to STRING or if Impala accepts TEXT as alias for STRING in its DDL (or if Impala ignores DDL constraints). Nevertheless, I think it's also reasonable to be more rigor on this end and leave stronger abstraction to tools like odo. That's why I submitted the issue here and not to impyla (but I CCed - and hereby again - impyla's maintainer @laserson in case he wants to pick it up).

> In any event, do you mind if I submit your fix upstream to impyla?

No, go ahead.

**Comment by laserson** [comment]
Just back from a vacation.  AFAIK, people generally use the DB API impl in impyla directly, rather than the SQLalchemy dialect, which I created primarily to work with Blaze, and which I generally based off of PyHive.  So I'm not opposed to a fix in impyla at the moment.
